onnxruntime/core/graph/graph.cc

```diff
@@ -380,13 +412,64 @@ common::Status NodeArg::UpdateTypeAndShape(const ONNX_NAMESPACE::TypeProto& inpu
           *current_type.mutable_sparse_tensor_type() = input_tensor_type;
         }
       }
-
       break;
     }
 #endif
+
+    case TypeProto::kOptionalType: {
+      if (utils::HasOptionalTensorType(input_type) && !utils::HasOptionalTensorType(current_type)) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Optional Type mismatch. Expected: Seq tensor. Got: Tensor");
+      } else if (!utils::HasOptionalTensorType(input_type) && utils::HasOptionalTensorType(current_type)) {
```

Comment:
Is this message correct?

Reply:
Fixed. You're right - combined the 2 checks and just have one ORT_MAKE_STATUS now

Reply:
I also made a helper for the logic that overrides types . This piece of code is common for tensors, sparse tensors, and optional tensors. So, that should be some binary savings.